### PR TITLE
Disable PGO instrumentation on naked function

### LINKED ISSUE #74573: Should `naked` imply `no_profile_instrument_func`?
functions marked as `naked` can only contain assembly code, and are in charge of handling the function prologues and epilogues. Under `-mllvm -pgo-temporal-instrumentation`, the compiler adds a call to `@llvm.instrprof.timestamp`, which can mess up with the assembly code in the function body.

What should we do then? Disable profiling of naked functions? Emit a warning / an error when enabling pgo temporal instrumentation ? Insert code while saving registers etc so that the call to the profiling function doesn't have any side effect?

### PR DESCRIPTION
**Pull request by serge-sans-paille**
We only allow for assembly code in naked function, and PGO instrumentation (esp. temporal instrumentation that introduces a function call) can wreak havoc in this.

Fix #74573

### CODE REVIEW

File: clang/lib/CodeGen/CodeGenFunction.cpp [hunk 1]
```diff
@@ -892,6 +892,10 @@ void CodeGenFunction::StartFunction(GlobalDecl GD, QualType RetTy,
     }
   }
 
+  if (FD->hasAttr<NakedAttr>()) {
```
Comment (teresajohnson) [inline clang/lib/CodeGen/CodeGenFunction.cpp:895]: Is this change needed given the separate change in LLVM skipPGOGen?

File: llvm/lib/Transforms/Instrumentation/PGOInstrumentation.cpp [hunk 2]
```diff
@@ -1783,6 +1783,8 @@ static bool skipPGOUse(const Function &F) {
 static bool skipPGOGen(const Function &F) {
   if (skipPGOUse(F))
     return true;
+  if (F.hasFnAttribute(llvm::Attribute::Naked))
```
Comment (MaskRay) [inline llvm/lib/Transforms/Instrumentation/PGOInstrumentation.cpp:1786]: Makes sense. Similar to https://reviews.llvm.org/D149721

File: clang/lib/CodeGen/CodeGenFunction.cpp (hunk 1 above)
Comment (MaskRay) [reply clang/lib/CodeGen/CodeGenFunction.cpp:895]: This should not be needed. The llvm change is sufficient to work with all frontends including clang.

**Review by MaskRay — APPROVED** [review]

File: clang/lib/CodeGen/CodeGenFunction.cpp (hunk 1 above)
Comment (serge-sans-paille) [reply clang/lib/CodeGen/CodeGenFunction.cpp:895]: not at all, that was a leftover, sorry about that.

**Review by teresajohnson — APPROVED** [review]

**Comment by serge-sans-paille** [comment]
Thanks @MaskRay & @teresajohnson for the quick review o/
